jetty-core/jetty-websocket/jetty-websocket-core-common/src/main/java/org/eclipse/jetty/websocket/core/FrameHandler.java

```diff
@@ -16,102 +16,112 @@
+ * <li>The {@link #onClosed(CloseStatus, Callback)} method is always
+ * called once a WebSocket connection is terminated, either gracefully
+ * or not. The error code will indicate the nature of the close.</li>
  * </ul>
+ * <p>FrameHandler is responsible to manage the demand for more
+ * WebSocket frames, either directly by calling {@link CoreSession#demand(long)}
+ * or by delegating the demand management to other components.</p>
  */
 public interface FrameHandler extends IncomingFrames
 {
     /**
-     * Async notification that Connection is being opened.
-     * <p>
-     * FrameHandler can write during this call, but can not receive frames until the callback is succeeded.
-     * </p>
-     * <p>
-     * If the FrameHandler succeeds the callback we transition to OPEN state and can now receive frames if auto-demanding,
-     * or can now call {@link CoreSession#demand(long)} to receive frames if it is not auto-demanding.
-     * If the FrameHandler fails the callback a close frame will be sent with {@link CloseStatus#SERVER_ERROR} and
-     * the connection will be closed. <br>
-     * </p>
+     * <p>Invoked when the WebSocket connection is opened.</p>
+     * <p>It is allowed to send WebSocket frames via
+     * {@link CoreSession#sendFrame(Frame, Callback, boolean)}.
+     * <p>WebSocket frames cannot be received  until a call to
+     * {@link CoreSession#demand(long)} is made.</p>
+     * <p>If the callback argument is failed, the implementation
+     * sends a CLOSE frame with {@link CloseStatus#SERVER_ERROR},
+     * and the connection will be closed.</p>
      *
      * @param coreSession the session associated with this connection.
-     * @param callback the callback to indicate success in processing (or failure)
+     * @param callback the callback to indicate success or failure of
+     * the processing of this event.
      */
     void onOpen(CoreSession coreSession, Callback callback);
 
     /**
-     * Receiver of all Frames.
-     * This method will never be called in parallel for the same session and will be called
-     * sequentially to satisfy all outstanding demand signaled by calls to
-     * {@link CoreSession#demand(long)}.
-     * Control and Data frames are passed to this method.
-     * Close frames may be responded to by the handler, but if an appropriate close response is not
-     * sent once the callback is succeeded, then a response close will be generated and sent.
+     * <p>Invoked when a WebSocket frame is received.</p>
+     * <p>This method will never be called concurrently for the
+     * same session; will be called sequentially to satisfy the
+     * outstanding demand signaled by calls to
+     * {@link CoreSession#demand(long)}.</p>
+     * <p>Both control and data frames are passed to this method.</p>
+     * <p>CLOSE frames may be responded from this method, but if
+     * they are not responded, then the implementation will respond.
```

Comment:
How does the implementation know that the CLOSE will not be responded to?
Should the signature be like our handle method:

```java
    boolean onFrame(Frame frame, Callback callback);
```

If true is returned then the application will invoke the callback, if false is returned then the application will handle the frame (pong for ping, close for close etc.).

Reply:
As per our call, the close logic is done when the callback is completed, not when the method returns.  That is OK.... but needs javadoc.